### PR TITLE
Make ContentDialog.ShowAsync awaitable return result on Hide() call

### LINKED ISSUE #4428: [iOS][Android] await ContentDialog.ShowAsync() doesn't return after Hide() is called
## Current behavior

When hiding a shown `ContentDialog cd` with `cd.Hide()`, the `await cd.ShowAsync()` method doesn't return. I have verified this on Android and iOS.

## Expected behavior

After hiding the `ContentDialog cd` with `cd.Hide()`, the `cd.ShowAsync()` returns, allowing the program to continue execution. This works on UWP.

## How to reproduce it (as minimally and precisely as possible)

Repro project located at: https://github.com/pdecostervgls/uno-repro-contentdialogclosing

## Workaround

None known.

## Environment

Nuget Package:

- [x] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s):
Uno.UI 3.2.0-dev.377

Affected platform(s):

- [x] iOS
- [x] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [x] Visual Studio 2019 (version: 16.7.6)
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

Relevant plugins:

- [ ] Resharper (version: )

### PR DESCRIPTION
**Pull request by pdecostervgls**
GitHub Issue (If applicable): fixes #4428

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

On Android and iOS, awaiting `ContentDialog.ShowAsync()` didn't return with `ContentDialogResult.None` if `ContentDialog.Hide()` was called to dismiss it.

## What is the new behavior?

It now does.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

I don't think this PR contains any breaking changes.

### CODE REVIEW

**Review by MartinZikmund — CHANGES_REQUESTED** [review]
It seems that one related test fails on WASM now `SamplesApp.UITests.Windows_UI_Xaml_Controls.ContentDialogTests.ContentDialog_Tests.ContentDialog_Closing_Deferred`. Can you check if the change has broken some behavior the test depends on?

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ContentDialogTests/ContentDialog_Closing.xaml.cs [hunk 1]
```diff
@@ -58,14 +58,14 @@ private void DeferredDialog_Click(object sender, RoutedEventArgs args)
 			var defer1Button = new _Button { Name = "Complete1Button", Content = "Complete 1" };
 			defer1Button.Click += (o, e) =>
 			{
-				deferral1.Complete();
 				ResultTextBlock.Text = "First complete called";
+				deferral1.Complete();
```
Comment (davidjohnoliver) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ContentDialogTests/ContentDialog_Closing.xaml.cs:62]: Why was this change needed?

**Review by davidjohnoliver — COMMENTED** [review]
Can you add a new UI test that verifies the sample you added?

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ContentDialogTests/ContentDialog_Closing.xaml.cs (hunk 1 above)
Comment (pdecostervgls) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ContentDialogTests/ContentDialog_Closing.xaml.cs:62]: I observed that when all deferral's were being completed in the Click handler, the execution of the click handler was stopped. Probably due to the Button going out of scope when the dialog got dismissed. This caused the TextBlock not being updated and the UI test failing. I wasn't completely sure what caused it, but I felt like the change was minor enough.

Comment (pdecostervgls) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ContentDialogTests/ContentDialog_Closing.xaml.cs:62]: > Can you add a new UI test that verifies the sample you added?
I could try to do that. I'm not very comfortable with the test framework yet though.

**Comment by pdecostervgls** [comment]
> Can you add a new UI test that verifies the sample you added?

I could try to do that. I'm not very comfortable with the test framework yet though.

**Review by davidjohnoliver — APPROVED** [review]
